[mlir][spirv] Implement UMod canonicalization for vector constants

fairywreath opened the pull request:
Closes #63174. 

Implements this transformation pattern, which is currently only applied to scalars, for vectors:
```
%1 = "spirv.UMod"(%0, %CONST_32) : (i32, i32) -> i32
%2 = "spirv.UMod"(%1, %CONST_4) : (i32, i32) -> i32
```
to
```
%1 = "spirv.UMod"(%0, %CONST_32) : (i32, i32) -> i32
%2 = "spirv.UMod"(%0, %CONST_4) : (i32, i32) -> i32
```

Additionally fixes and issue where patterns like this:
```
%1 = "spirv.UMod"(%0, %CONST_4) : (i32, i32) -> i32
%2 = "spirv.UMod"(%1, %CONST_32) : (i32, i32) -> i32
```
were incorrectly canonicalized to:
```
%1 = "spirv.UMod"(%0, %CONST_4) : (i32, i32) -> i32
%2 = "spirv.UMod"(%0, %CONST_32) : (i32, i32) -> i32
```
which is incorrect since `(X % A) % B`  == `(X % B)` IFF A is a multiple of B, i.e., B divides A.

fairywreath committed:
[mlir][spirv] Implement UMod canonicalization for vector constants

IgWod-IMG commented:
Me again :) I have two minor comments:

1) Could you please reference the issues you fix (https://github.com/llvm/llvm-project/issues/63174) in the description. In fact, if you use one of those [magic keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) GitHub should automatically close the issue, once this change is merged.

2) It'd probably be useful to include the pattern that was broken in the tests (if not already there), so it can catch any future regression.

fairywreath committed:
Add more test and improve code

fairywreath committed:
Add one more test

fairywreath commented:
@IgWod-IMG Thanks for the review. The vector test I added actually tested the pattern that was broken. Anyways I have added more tests, one which is a scalar version on the pattern that was broken.

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -336,19 +335,28 @@ struct UModSimplification final : OpRewritePattern<spirv::UModOp> {
+    // not, fail the transformation.
+    bool isApplicable = false;
+    if (auto prevInt = dyn_cast<IntegerAttr>(prevValue)) {
+      auto currInt = dyn_cast<IntegerAttr>(currValue);
```
Use `cast` to assert that the cast succeeded and the value is not null

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -336,19 +335,28 @@ struct UModSimplification final : OpRewritePattern<spirv::UModOp> {
+    } else if (auto prevVec = dyn_cast<DenseElementsAttr>(prevValue)) {
+      auto currVec = dyn_cast<DenseElementsAttr>(currValue);
+      isApplicable = llvm::all_of(
+          llvm::zip(prevVec.getValues<APInt>(), currVec.getValues<APInt>()),
```
can we use zip_equal?

kuhar commented on mlir/lib/Dialect/SPIRV/IR/SPIRVCanonicalization.cpp:
```diff
@@ -336,19 +335,28 @@ struct UModSimplification final : OpRewritePattern<spirv::UModOp> {
+      isApplicable = llvm::all_of(
+          llvm::zip(prevVec.getValues<APInt>(), currVec.getValues<APInt>()),
+          [](const auto &pair) {
+            return std::get<0>(pair).urem(std::get<1>(pair)) == 0;
```
nit: alternatively, can you use structured bindings and name these variables for better readability? Something like:

```c++
auto &[nameA, nameB] = pair;
return nameA.urem(nameB);
```

kuhar reviewed (commented):
Thanks for improving this

fairywreath committed:
Address review comments
